### ISSUE TITLE
Simplified Data Package Forms

### ISSUE THREAD

**Issue by p-a-s-c-a-l** [clarity-h2020/docker-drupal#78]
Implement simplified Data Package and Data Package Resource edit forms with minimal required properties.

**Comment by maesbri** [comment]
@patrickkaleta, as discussed in Vienna, we need to change the way of creating resources within the data package. Currently, there is only the possibility to add a new one (like this):

![image](https://user-images.githubusercontent.com/9677881/60799344-40ae8180-a173-11e9-812a-40fc95fcd8d3.png)

But it should be replaced (as in the "contributors" property), in order to allow the reuse of existing resources already available:

![image](https://user-images.githubusercontent.com/9677881/60799458-76ec0100-a173-11e9-8d3a-2342f497d3ec.png)

**Comment by maesbri** [comment]
Currently, the "tags" properties is displayed like this (it is a list where one or more values from various taxonomies can be selected), which is not very user-friendly right now (user must hold CTRL in order to select various values while clicking with the mouse, e.g., service-type:wfs, eu-gl:hazard-characterization, etc...).

![image](https://user-images.githubusercontent.com/9677881/60799844-435da680-a174-11e9-8f09-40419a3fdf2d.png)

@fgeyer16 , I think you mentioned in some previous teleconference that you would try to make it more user-friendly by displaying it in another way (although all the selected values must still be stored in "tags" properties). Do you think this feature in the form could be ready during this week?

**Comment by p-a-s-c-a-l** [comment]
> I think you mentioned in some previous teleconference that you would try to make it more user-friendly by displaying it in another way (although all the selected values must still be stored in "tags" properties).

IMHO this feature is described here: https://github.com/clarity-h2020/docker-drupal/issues/66

**Comment by maesbri** [comment]
@patrickkaleta , I have created a custom display settings (named SimplifiedDataPackage) with only the minimum list of properties we need for the DCs starting to create their data packages (this has been applied both for the dp_data_package and dp_resource entity types). However, I don't know how to use it, as when I try to create a new data package the default form still appears with all fields.

btw, in the resource form I would like to remove all tabs (as now the list of properties is much shorter). If I remove the tabs do they get destroyed for all form displays (i.e., the default one). I didn't test it as it is an operation that cannot be undone ...

**Comment by patrickkaleta** [comment]
> @patrickkaleta , I have created a custom display settings (named SimplifiedDataPackage) with only the minimum list of properties we need for the DCs starting to create their data packages (this has been applied both for the dp_data_package and dp_resource entity types). However, I don't know how to use it, as when I try to create a new data package the default form still appears with all fields.

This new form display `SimplifiedDataPackage` (btw: I would've chosen something more generic like "Simplified", since it can be used for all entity types independently now. But that's not an issue) has to be "called" specifically now for each content type that is using it (otherwise by default always the "default" form display will be used). How that is done exactly depends on the way those forms are being used.

- for the **datapackage** I believe they are being created via the [Content page](https://csis.myclimateservice.eu/admin/content), right? Here, instead of clicking on "+ Add content" the user has to click on "+ Add node as SimplifiedDataPackage". Then he gets an overview of all content types configured for the "SimplifiedDataPackage" form display, where he can chose dp_data_package.
**Note:** I could work on a designated button for the User's profile page, that would automatically go to the Simplified version of the form. I can have that done by the end of the week.

- for the **resource** the user can do simply follow the same approach as for the data package if he likes to create the resources first without the context of a data package (once he is done with all the resources, he can then create the data package and  just select the necessary resources), or he creates them during the creation the data package. For that the corresponding entity browser has to be adapted [here ](https://csis.myclimateservice.eu/admin/config/content/entity_browser/data_resources/widgets) (Form mode = SimplifiedDataPackage) to display the simplified form. I've already done that now.

If you want to display the simplified form anywhere else, please let me know.


> btw, in the resource form I would like to remove all tabs (as now the list of properties is much shorter). If I remove the tabs do they get destroyed for all form displays (i.e., the default one). I didn't test it as it is an operation that cannot be undone ...

It is safe to remove those tabs, since that only affects the current selected form display. But it's not necessary, if those tabs have no fields assigned than they are not displayed in the form (that's what I did now for the simplified form of the Resource).

> @fgeyer16 , I think you mentioned in some previous teleconference that you would try to make it more user-friendly by displaying it in another way (although all the selected values must still be stored in "tags" properties). Do you think this feature in the form could be ready during this week?

as mentioned by @p-a-s-c-a-l, #66 could be used to improve this. I will have a look at it now and let you know once I made the necessary adjustments. Please have then a look at it and tell me if all needed tags are still available in there. 

@maesbri if you like we can have a telco regarding those forms this week. I'm available all week expect for Wednesday morning.

**Comment by patrickkaleta** [comment]
**Update:**
On the user's profile page (.../user/[userID]) there is now a new tab (only visible for Admins, Developers and Editors) labeled "My content". On this tab a list of all data packages and resources created by this user is shown. 
Additionally, two buttons for adding Data packages and Resources using the new simplified form are there as well.

**Comment by patrickkaleta** [comment]
> Currently, the "tags" properties is displayed like this (it is a list where one or more values from various taxonomies can be selected), which is not very user-friendly right now (user must hold CTRL in order to select various values while clicking with the mouse, e.g., service-type:wfs, eu-gl:hazard-characterization, etc...).
> 
> ![image](https://user-images.githubusercontent.com/9677881/60799844-435da680-a174-11e9-8f09-40419a3fdf2d.png)
> 
> @fgeyer16 , I think you mentioned in some previous teleconference that you would try to make it more user-friendly by displaying it in another way (although all the selected values must still be stored in "tags" properties). Do you think this feature in the form could be ready during this week?

@maesbri: I configured the [entity queue](https://csis.myclimateservice.eu/admin/structure/entityqueue/dp_tags/list) for the `tags_field`, so please have a look at the simplified form and let me know if all necessary items are being displayed in the form field now. I'm still a little confused when it comes to which hazards should be displayed in the forms. I followed this approach:

**Example for Hazards taxonomy:**
-temperature (NO - should not be an available option, since it's a root term and subterms are available)
--heat (YES - should be available)
---extreme-heat (YES - should probably also available?)
----hot-days-75p (NO - because it's the index)
...
--cold (YES - just like "heat")
...
-drought (YES - although it's a root term, it should be displayed since no subterms available)
--consecutive-dry-days (NO - because it's an index)

**Comment by patrickkaleta** [comment]
> @patrickkaleta, as discussed in Vienna, we need to change the way of creating resources within the data package. Currently, there is only the possibility to add a new one (like this):
> 
> ![image](https://user-images.githubusercontent.com/9677881/60799344-40ae8180-a173-11e9-812a-40fc95fcd8d3.png)
> 
> But it should be replaced (as in the "contributors" property), in order to allow the reuse of existing resources already available:
> 
> ![image](https://user-images.githubusercontent.com/9677881/60799458-76ec0100-a173-11e9-8d3a-2342f497d3ec.png)

@maesbri: 
**Done!** Let me know if you would like to have additional adjustments (display some other fields, add more filters, etc).

**Comment by maesbri** [comment]
> @maesbri:
> **Done!** Let me know if you would like to have additional adjustments (display some other fields, add more filters, etc).

Yes, it works. Thank you!

**Comment by maesbri** [comment]
> @maesbri: I configured the [entity queue](https://csis.myclimateservice.eu/admin/structure/entityqueue/dp_tags/list) for the `tags_field`, so please have a look at the simplified form and let me know if all necessary items are being displayed in the form field now. I'm still a little confused when it comes to which hazards should be displayed in the forms. I followed this approach:
> 
> **Example for Hazards taxonomy:**
> -temperature (NO - should not be an available option, since it's a root term and subterms are available)
> --heat (YES - should be available)
> ---extreme-heat (YES - should probably also available?)
> ----hot-days-75p (NO - because it's the index)
> ...
> --cold (YES - just like "heat")
> ...
> -drought (YES - although it's a root term, it should be displayed since no subterms available)
> --consecutive-dry-days (NO - because it's an index)

That' fine, but there is a problem now in the tags list. I cannot select any hazard-index or a vulnerability class (as they are sub-terms in the hazard and elements_at_risk taxonomies. Separating hazards and elements at risk from hazard indexes and vulnerability classes from elements at risk it is better for the end user, but is there a way to do display it in the list?). Bwt, is there any possibility to make the tag list vertically larger so the enduser has not to continuosly scroll down?

**Comment by maesbri** [comment]
> **Update:**
> On the user's profile page (.../user/[userID]) there is now a new tab (only visible for Admins, Developers and Editors) labeled "My content". On this tab a list of all data packages and resources created by this user is shown.
> Additionally, two buttons for adding Data packages and Resources using the new simplified form are there as well.

Thanks, I see now them :)

**Comment by maesbri** [comment]
@p-a-s-c-a-l , @patrickkaleta 
Now the id property (in data-package and resource) is hidden and filled in automatically with the [node:uuid], so it is transparent for the end-user.

**Update:**  This works for the new data-packages and resources but how do we replace the old id values in the existing resources/data-packages with the node uuid?

**Comment by maesbri** [comment]
@p-a-s-c-a-l , @DenoBeno 
According to our discussion yesterday, I have updated the simplified data-package creation form by adding the following properties at:
- data-package level: contributors and licenses
- at resource level: sources, contributors and licenses

**Comment by patrickkaleta** [comment]
> @p-a-s-c-a-l , @patrickkaleta
> Now the id property (in data-package and resource) is hidden and filled in automatically with the [node:uuid], so it is transparent for the end-user.
> 
> **Update:** This works for the new data-packages and resources but how do we replace the old id values in the existing resources/data-packages with the node uuid?

I think since there are currently still not that many datapackages/resources, the easiest way would be to **edit and save without changes** each of the existing DPs and resources. The ID will be updated automatically thanks to your settings.

**Comment by patrickkaleta** [comment]
> > @maesbri: I configured the [entity queue](https://csis.myclimateservice.eu/admin/structure/entityqueue/dp_tags/list) for the `tags_field`, so please have a look at the simplified form and let me know if all necessary items are being displayed in the form field now. I'm still a little confused when it comes to which hazards should be displayed in the forms. I followed this approach:
> > **Example for Hazards taxonomy:**
> > -temperature (NO - should not be an available option, since it's a root term and subterms are available)
> > --heat (YES - should be available)
> > ---extreme-heat (YES - should probably also available?)
> > ----hot-days-75p (NO - because it's the index)
> > ...
> > --cold (YES - just like "heat")
> > ...
> > -drought (YES - although it's a root term, it should be displayed since no subterms available)
> > --consecutive-dry-days (NO - because it's an index)
> 
> That' fine, but there is a problem now in the tags list. I cannot select any hazard-index or a vulnerability class (as they are sub-terms in the hazard and elements_at_risk taxonomies. Separating hazards and elements at risk from hazard indexes and vulnerability classes from elements at risk it is better for the end user, but is there a way to do display it in the list?). Bwt, is there any possibility to make the tag list vertically larger so the enduser has not to continuosly scroll down?

I understand what you'd like to see. Not sure if it will be possible to separate e.g. hazards from hazard indexes, since they are stored in the same taxonomy. I will give it a quick try, otherwise they will remain together.

Regarding the height of the select box, this is something that should be set with CSS in our new theme.

**Comment by DenoBeno** [comment]
I have implemented several changes to the way data packages are defined and to the way they are presented.

Fro the start, I have made a DP presentation somewhat nicer. Teh result can be seen e.g. on https://csis.myclimateservice.eu/node/221. Screenshot is a bit too big, so let me explain:

1. separation of content into "data package summary" and "ressources" section
2. Most of the resource description is hidden under "Details"
![grafik](https://user-images.githubusercontent.com/2961790/61051721-0569b800-a3ea-11e9-9c91-5af389468fa8.png)

3. Two fields are calculated automatically: Resource contributors and Ressource licenses (naming convention tbd.

![grafik](https://user-images.githubusercontent.com/2961790/61051844-5aa5c980-a3ea-11e9-9647-ac0deafa9049.png)

![grafik](https://user-images.githubusercontent.com/2961790/61051865-698c7c00-a3ea-11e9-8434-dcc38a7949bd.png)

This is performed by means of two Entity View Atachement views:

- https://csis.myclimateservice.eu/admin/structure/views/view/dp_contributors_eva?destination=/admin/structure/views
- https://csis.myclimateservice.eu/admin/structure/views/view/dp_license_eva?destination=/admin/structure/views

This means that:

1. There is no reason to list all resource contributors in the DP contributors field. We can easily calculate this on the fly.
2. All licenses must have a "license type" field (implemented) pointing to DP license type taxonomy https://csis.myclimateservice.eu/admin/structure/taxonomy/manage/dp_licensetype/overview

Currently defined license types are: 
![grafik](https://user-images.githubusercontent.com/2961790/61052078-eae40e80-a3ea-11e9-958b-1aba90faa607.png)

In addition, I have also disabled editing of the contributors and resources within data package edit form (see e.g. https://csis.myclimateservice.eu/node/221/edit/simplifieddatapackage) and added the "licenses" to resource "short" display. Result looks like this:
![grafik](https://user-images.githubusercontent.com/2961790/61052651-26330d00-a3ec-11e9-966a-1b4e98bef771.png)

As you can see, we haven't defined the licenses for most (all?) of the ressources. Btw, I wonder if we will ever have more than one license per resource?

**Comment by DenoBeno** [comment]
What is missing now?

- first, we should IMO allow the normal users to ONLY use the simplified edit forms. Alternatively, we could simplify default forms and allow only special users to use the cmplex one.
- second, the users need a simple way to start editing already existign DPs and adding new ones. 
- while resources and contributors can be easily added from DP, it's not so easy to change them afterwards. I removed edit buttons in the form because they were leading to standard edit form and not to simplified one but we need soem way to handle this.
- third, we need some inline explanations (help) for the users how to fill in the fields.
- Finally, we need to think again why we need three title fields for each entity. If you need to generate a string where spaces are replaced by dashes, this can eb done automatically using a view. If you need unique ID, this can be generated automatically by teh system too (UUID is geenrated automaticaly AFAIK). So why do we complicate users' lives?

I guess there are other issues too, but this comes to mind now.

Ah yes: it would be even easier to sum up all the licenses used for ressources and show them. In fact, that's exactly the same as listing all contributors on a technical levels. we can have one or another or both - it's up to us to decide. In any case, there is no need to repeat the listing of the underlying licenses and contributors manually at the DP level.

**Comment by maesbri** [comment]
> > @p-a-s-c-a-l , @patrickkaleta
> > Now the id property (in data-package and resource) is hidden and filled in automatically with the [node:uuid], so it is transparent for the end-user.
> > **Update:** This works for the new data-packages and resources but how do we replace the old id values in the existing resources/data-packages with the node uuid?
> 
> I think since there are currently still not that many datapackages/resources, the easiest way would be to **edit and save without changes** each of the existing DPs and resources. The ID will be updated automatically thanks to your settings.

Yes!, I noticed while saving some of the resources :). I will do it with all the other resources once the old properties are removed, otherwise it will create again all the empty contents

**Comment by maesbri** [comment]
> As you can see, we haven't defined the licenses for most (all?) of the ressources. Btw, I wonder if we will ever have more than one license per resource?

This information is not available for all datasets. I did my best when started preparing the data package some months ago.
Btw, I opened in edit mode the agricultural areas resource, but in the licenses list it doesn't appear the ones that you added to the taxonomy.

![image](https://user-images.githubusercontent.com/9677881/61055002-df93e180-a3f0-11e9-95bb-8dbd3bbd8b48.png)

**Comment by p-a-s-c-a-l** [comment]
> This information is not available for all datasets.

For each resource in the Data Package, there *should* be a corresponding entry in our [CKAN](https://ckan.myclimateservice.eu/) where the license field is mandatory.

**Comment by patrickkaleta** [comment]
> Btw, I opened in edit mode the agricultural areas resource, but in the licenses list it doesn't appear the ones that you added to the taxonomy.
> 
> ![image](https://user-images.githubusercontent.com/9677881/61055002-df93e180-a3f0-11e9-95bb-8dbd3bbd8b48.png)

@maesbri 
The taxonomy "DP License Type" doesn't store actual licenses but only the different types. It groups the individual licenses (which are modelled as content type "DP Lincense") based on how restrictive they are.

So what changed is that the content type "DP lincense" now has a new field called `License type` (reference to the new License type taxonomy). That field is needed to aggregate all resource licenses of a data package and select the most restrictive one, which is then used as the overall license for the data package.

When creating a new resource you still only select a license and not directly a license type (that's what you see on the screenshot).

**Comment by maesbri** [comment]
> What is missing now?
> 
> * first, we should IMO allow the normal users to ONLY use the simplified edit forms. Alternatively, we could simplify default forms and allow only special users to use the cmplex one.
> * second, the users need a simple way to start editing already existign DPs and adding new ones.
> * while resources and contributors can be easily added from DP, it's not so easy to change them afterwards. I removed edit buttons in the form because they were leading to standard edit form and not to simplified one but we need soem way to handle this.
> * third, we need some inline explanations (help) for the users how to fill in the fields.

Explanations on how to produce datapackage/resources will be in the wiki page I will prepare. Several properties in the form already contain inline explanations, however it has its formatting limitations, which is not very convenient when having long explanations (e.g., it is not possible to have new lines within the explanatory text or making some words bold).

> * Finally, we need to think again why we need three title fields for each entity. If you need to generate a string where spaces are replaced by dashes, this can eb done automatically using a view. If you need unique ID, this can be generated automatically by teh system too (UUID is geenrated automaticaly AFAIK). So why do we complicate users' lives?

The unique ID is already generated and assigned to the "id" property. "name" is a human readable identifier for the resource that doesn't contain any strange characters so it doesn't cause problems for software managing it (I guess the idea behind was also that could be displayed for humans). Title and description are the long and longer (respectively) text free explanation of the resource. If you still consider annoying for the end-user to type the "name", then is there a way to get the "title", replace empty and strange characters with "-" and then assign it to the "name" property?
> 
> I guess there are other issues too, but this comes to mind now.
> 
> Ah yes: it would be even easier to sum up all the licenses used for ressources and show them. In fact, that's exactly the same as listing all contributors on a technical levels. we can have one or another or both - it's up to us to decide. In any case, there is no need to repeat the listing of the underlying licenses and contributors manually at the DP level.

I think it is not that trivial with the license. Each resource (may) has its license, but this doesn't mean that the license of the data package is the sum of all of them. To me it makes more sense, that the user gives a license to the data package that it is compatible with all the underlying licenses. However, implementing such a feature is not trivial at all as you need to know the compatibility among licenses (e.g., (in a very simplified manner) all resources are LGPL but one with GPL license, then the only compatible license for the whole data package would be GPL)

**Update**: ok, seeing @patrickkaleta 's explanation about the license type I think this could do it :)

**Comment by maesbri** [comment]
> > This information is not available for all datasets.
> 
> For each resource in the Data Package, there _should_ be a corresponding entry in our [CKAN](https://ckan.myclimateservice.eu/) where the license field is mandatory.

I think that would not be enough, someone has to manually add the license to each resource (in drupal) so it can be later on used to restrict the data package license (or even visualize the license in the CSIS). Otherwise, the alternative would be having to follow the metadata link of the resource, parse the json provided by ckan, extract the license information and then display it.

**Comment by patrickkaleta** [comment]
> > @maesbri: I configured the [entity queue](https://csis.myclimateservice.eu/admin/structure/entityqueue/dp_tags/list) for the `tags_field`, so please have a look at the simplified form and let me know if all necessary items are being displayed in the form field now. I'm still a little confused when it comes to which hazards should be displayed in the forms. I followed this approach:
> > **Example for Hazards taxonomy:**
> > -temperature (NO - should not be an available option, since it's a root term and subterms are available)
> > --heat (YES - should be available)
> > ---extreme-heat (YES - should probably also available?)
> > ----hot-days-75p (NO - because it's the index)
> > ...
> > --cold (YES - just like "heat")
> > ...
> > -drought (YES - although it's a root term, it should be displayed since no subterms available)
> > --consecutive-dry-days (NO - because it's an index)
> 
> That' fine, but there is a problem now in the tags list. I cannot select any hazard-index or a vulnerability class (as they are sub-terms in the hazard and elements_at_risk taxonomies. Separating hazards and elements at risk from hazard indexes and vulnerability classes from elements at risk it is better for the end user, but is there a way to do display it in the list?). Bwt, is there any possibility to make the tag list vertically larger so the enduser has not to continuosly scroll down?

@maesbri 
it took me a while but I think I found a reasonable solution. Using EntityQueue on its own cannot provide the possibility to show terms from the same taxonomy under different labels (like it would be necessary for hazards and hazard indexes or El. at risk and vulnerability classes). Also there are now already roughly 100 tags, so using a simple select list is not an optimal solution.

So now the [Entityqueue](https://www.drupal.org/project/entityqueue) module is working together with the [Entity Browser](https://www.drupal.org/project/entity_browser) module (which we already use for selecting sources and contributors) and the result looks like this:
![tag-selection-entity-browser](https://user-images.githubusercontent.com/42167036/61124656-b2eed100-a4a7-11e9-9adc-65b39bbd730a.png)

Tags are separated in tabs by their type and the user just clicks on the items he would like to add and at the bottom of the table the current selection is shown. Visual appearance can be further improved once the new theme is active.

Please have a look if now all tags are available or if there is still something missing.

**Comment by maesbri** [comment]
> Tags are separated in tabs by their type and the user just clicks on the items he would like to add and at the bottom of the table the current selection is shown. Visual appearance can be further improved once the new theme is active.
> 
> Please have a look if now all tags are available or if there is still something missing.

Thank you very much @patrickkaleta  . I think it is quite a good solution :)

**Comment by patrickkaleta** [comment]
Just noticed that I forgot about the damage classes... I added them to the tags selection as well.

**Comment by p-a-s-c-a-l** [comment]
> I think that would not be enough, someone has to manually add the license to each resource (in drupal) so it can be later on used to restrict the data package license (or even visualize the license in the CSIS).

Yes, that's sufficient. Keep it simple. :-)

**Comment by DenoBeno** [comment]
Few updates on dp_resource node type.

1. field_name has a "CKAN ID" label in cp_resource now. We can try to use it for calling the CKAN API at some later point. (Parsing the license field on CKAN to fill in some fields would be possible in theory, but I don't know how to do it and I have no idea when I'll have time to play with that. Or anyone else... So we need to enter them per hand for now. We can have a default one for the new resources...)
2. field_title is automatically filled in with the content of title field now, so the user doesn't have to copy teh title to this field. We even automatically append "OBSOLETE" to it, so that we can easily see which widgets need to be updated. 
3. I have changed the Default/compact/short description and teaser view modes of dp_resource to show both the title and field_title for now. Once the field_title is deleted, it will disapear on its own. (I see that Miguel has removed them now - not sure if that's a good idea, since we need to know the old name in order to rename the resources)
4. I have also edited both form displays to only show the title field. The user now sees just this one field.
